src/libraries/System.Linq/src/System/Linq/Shuffle.SpeedOpt.cs

```diff
@@ -0,0 +1,272 @@
+                            reservoir.Add(list[i]);
+                        }
+                        else
+                        {
+                            // For each subsequent element in the source, randomly replace an element in the
+                            // reservoir with a decreasing probability.
+                            int r = Random.Shared.Next(i + 1);
+                            if (r < takeCount)
+                            {
+                                reservoir[r] = list[i];
+                            }
+                        }
+                    }
+                }
+                else
+                {
+                    using IEnumerator<TSource> e = source.GetEnumerator();
+                    if (e.MoveNext())
+                    {
+                        reservoir = [e.Current];
+
+                        // Fill the reservoir with the first takeCount elements from the source.
+                        // If we can't fill it, just return what we get.
+                        while (reservoir.Count < takeCount)
+                        {
+                            if (!e.MoveNext())
+                            {
+                                goto ReturnReservoir;
+                            }
+
+                            reservoir.Add(e.Current);
+                        }
+
+                        // For each subsequent element in the source, randomly replace an element in the
+                        // reservoir with a decreasing probability.
+                        long totalElementsSeen = reservoir.Count;
+                        while (e.MoveNext())
+                        {
+                            long r = Random.Shared.NextInt64(totalElementsSeen);
+                            if (r < reservoir.Count)
+                            {
+                                reservoir[(int)r] = e.Current;
+                            }
+                        }
+                    }
+                }
+
+            ReturnReservoir:
+                if (reservoir is not null)
+                {
+                    // Ensure that elements in the reservoir are in random order. The sampling helped
+                    // to ensure we got a uniform distribution from the source into the reservoir, but
+                    // it didn't randomize the order of the reservoir itself; this is especially relevant
+                    // to the elements initially added into the reservoir.
+                    Random.Shared.Shuffle(CollectionsMarshal.AsSpan(reservoir));
+                }
+
+                return reservoir;
+            }
+
+            public override TSource[] ToArray() => SampleToList(_source, _takeCount)?.ToArray() ?? [];
+
+            public override List<TSource> ToList() => SampleToList(_source, _takeCount) ?? [];
+
+            public override int GetCount(bool onlyIfCheap) =>
+                !onlyIfCheap ? Math.Min(_takeCount, _source.Count()) :
```

Comment:
`_source` isn't a take-limited enumerator, is it?  Would we want this to be `Min(_takeCount, _source.Take(_takeCount).Count())`, or would under-enumerating there be a Linq side-effect no-no?

Reply:
Oh, interesting. The downside to that is allocating another iterator object for the Take. The upside is we could avoid a ton of enumeration. On the balance, it's probably worthwhile. I've made the change but swapped the order of the checks so we only take that path if we can't get the count O(1).